pkg/daemon/rpm-ostree.go

```diff
@@ -321,51 +324,90 @@ func (r *RpmOstreeClient) Rebase(imgURL, osImageContentDir string) (changed bool
 	return
 }
 
-// Rebase potentially rebases system if not already rebased.
-func (r *RpmOstreeClient) RebaseLayered(imgURL string) (changed bool, err error) {
-
-	defaultDeployment, err := r.GetBootedDeployment()
-	if err != nil {
-		return
-	}
-
-	previousPivot := ""
-	if len(defaultDeployment.CustomOrigin) > 0 {
-		if strings.HasPrefix(defaultDeployment.CustomOrigin[0], "pivot://") {
-			previousPivot = defaultDeployment.CustomOrigin[0][len("pivot://"):]
-			glog.Infof("Previous pivot: %s", previousPivot)
-		} else {
-			glog.Infof("Previous custom origin: %s", defaultDeployment.CustomOrigin[0])
-		}
-	} else {
-		glog.Info("Current origin is not custom")
-	}
-
+// RebaseLayered rebases system or errors if already rebased
+func (r *RpmOstreeClient) RebaseLayered(imgURL string) (err error) {
 	glog.Infof("Executing rebase to %s", imgURL)
 	args := []string{"rebase", "--experimental", "ostree-unverified-registry:" + imgURL}
 
-	if err = runRpmOstree(args...); err != nil {
-		return
-	}
-
-	changed = true
-	return
+	return runRpmOstree(args...)
 }
 
 // Live apply live-applies whatever we rebased to
 func (r *RpmOstreeClient) ApplyLive() (err error) {
 
 	glog.Infof("Applying live")
 
-	args := []string{"ex", "apply-live", "--allow-replacement"}
+	return runRpmOstree("ex", "apply-live", "--allow-replacement")
+}
 
-	if err = runRpmOstree(args...); err != nil {
-		return
+type FileTree struct {
+	children map[string]*FileTree
+}
+
+// ensures nodes exist for every element in a path
+func (node *FileTree) insert(path []string) {
+	child, ok := node.children[path[0]]
+	if !ok {
+		// if we're inserting the first child for this node we need to create the map
+		if node.children == nil {
+			node.children = make(map[string]*FileTree)
+		}
+		// there isn't a node for the first element in the path, so we need to create one
+		child = &FileTree{}
+		node.children[path[0]] = child
 	}
+	if len(path) > 1 {
+		child.insert(path[1:])
+	}
+}
 
+// walk the tree, generating a list of all paths to leaves
+func (node *FileTree) walk() (paths []string) {
```

Comment:
It might just be the way my brain works, but this totally baked my noodle. I looked at it for way too long. :laughing: 

Am I understanding this right that the reason you do all this work parsing the diff into a FileTree is because: 
 - ostree diff includes the intermediate directories also 
 - you don't know which ones are the intermediate directories until you find the "leaves" ? 
 - once you find the leaves, you can figure out which ones are "extra" (non-leaf) and exclude them from the changeset?

Reply:
> It might just be the way my brain works, but this totally baked my noodle. I looked at it for way too long. laughing

Yeah see my comment above about overkill :grimacing: 

> 
> Am I understanding this right that the reason you do all this work parsing the diff into a FileTree is because:
> 
>   * ostree diff includes the intermediate directories also

Yep
> 
>   * you don't know which ones are the intermediate directories until you find the "leaves" ?

Right. We know a directory is not intermediate if it has no children or equivalently `len(pathsToLeaves) == 0`
> 
>   * once you find the leaves, you can figure out which ones are "extra" (non-leaf) and exclude them from the changeset?

Essentially. The recursion will result in going from children up to the root, so you don't exclude so much as prepend to an already existing path. E.g. if we have `/path/to/file`, when you're processing `to` you'll see that you have child `file` and put `/to/file` in paths. And then the same will happen to prepend `path`

Any ideas on making this more legible? Or taking a completely different approach? I'm not even sure if `ostree` outputting every directory is a bug or a feature. @cgwalters should ostree be outputting
```
/path
/path/to
/path/to/file
```